OpenRA.Mods.Common/Traits/Render/WithDecorationBase.cs

```diff
@@ -0,0 +1,147 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Support;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits.Render
+{
+	public enum BlinkState { Off, On }
+
+	public abstract class WithDecorationBaseInfo : ConditionalTraitInfo
+	{
+		[Desc("Position in the actor's selection box to draw the decoration.")]
+		public readonly DecorationPosition Position = DecorationPosition.TopLeft;
+
+		[Desc("Player stances who can view the decoration.")]
+		public readonly Stance ValidStances = Stance.Ally;
+
+		[Desc("Should this be visible only when selected?")]
+		public readonly bool RequiresSelection = false;
+
+		[Desc("Offset sprite center position from the selection box edge.")]
+		public readonly int2 Margin = int2.Zero;
+
+		[Desc("Screen-space offsets to apply when defined conditions are enabled.",
+			"A dictionary of [condition string]: [x, y offset].")]
+		public readonly Dictionary<BooleanExpression, int2> Offsets = new Dictionary<BooleanExpression, int2>();
+
+		[Desc("The number of ticks that each step in the blink pattern in active.")]
+		public readonly int BlinkInterval = 5;
+
+		[Desc("A pattern of ticks (BlinkInterval long) where the decoration is visible or hidden.")]
+		public readonly BlinkState[] BlinkPattern = { };
+
+		[Desc("Override blink conditions to use when defined conditions are enabled.",
+			"A dictionary of [condition string]: [pattern].")]
+		public readonly Dictionary<BooleanExpression, BlinkState[]> BlinkPatterns = new Dictionary<BooleanExpression, BlinkState[]>();
+
+		[ConsumedConditionReference]
+		public IEnumerable<string> ConsumedConditions
+		{
+			get { return Offsets.Keys.Concat(BlinkPatterns.Keys).SelectMany(r => r.Variables).Distinct(); }
+		}
+	}
+
+	public abstract class WithDecorationBase : ConditionalTrait<WithDecorationBaseInfo>, IDecoration
```

Comment:
Duplicate references to trait info in sub-classes can be removed by continuing the curiously recurring template pattern I gave `ConditionalTrait<Info>`
```c#
public abstract class WithDecorationBase<InfoType> : ConditionalTrait<Info>, IDecoration where InfoType : WithDecorationBaseInfo
```